File: neetcode/blind75/arrays_and_hashing/products_of_array_discluding_self.py
# https://neetcode.io/problems/products-of-array-discluding-self

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        if len(nums) == 0:
            return nums

        output = [1] * len(nums)

        for i in range(1, len(nums)):
            output[i] = output[i-1] * nums[i-1]

        postfix_sum = 1
        for i in range(len(nums)-1, -1, -1):
            output[i] *= postfix_sum
            postfix_sum *= nums[i]
        
        return output

        # left = [1] * len(nums)
        # right = [1] * len(nums)

        # for i in range(1, len(nums)):
        #     left[i] = left[i-1] * nums[i-1]

        # for i in range(len(nums)-2, -1, -1):
        #     right[i] = right[i+1] * nums[i+1]

        # for i in range(len(nums)):
        #     output[i] = right[i] * left[i]

        # return output

File: neetcode/blind75_JS/Arrays-and-hashing/products-of-array-discluding-self.js
// https://neetcode.io/problems/products-of-array-discluding-self

class Solution {
    /**
     * @param {number[]} nums
     * @return {number[]}
     */
    productExceptSelf(nums) {
        if (nums.length === 0) {
            return 0
        }

        // without division and O(n) time
        const leng = nums.length
        const left = Array(leng).fill(1)
        const right = Array(leng).fill(1)

        for (let i = 1; i < leng; i++) {
            left[i] = left[i-1] * nums[i-1]
        }

        for (let i = leng-2; i >= 0; i--) {
            right[i] = right[i+1] * nums[i+1]
        }

        for (let i = 0; i < leng; i++) {
            nums[i] = left[i] * right[i]
        }

        return nums
    }
}
